library/core/src/any.rs

```diff
@@ -202,10 +203,69 @@ pub trait Any: 'static {
 #[stable(feature = "rust1", since = "1.0.0")]
 impl<T: 'static + ?Sized> Any for T {
     fn type_id(&self) -> TypeId {
+        unsafe impl<T: ?Sized + Any> OverrideAnyTypeId for T {
+            default fn overridden_type_id() -> TypeId {
+                TypeId::of::<Self>()
+            }
+        }
+
+        <T as OverrideAnyTypeId>::overridden_type_id()
+    }
+}
+
+// Safety: overriding the dynamic `type_id` of a type must not be done
+// in way where arbitrary code may witness the static and dynamic type mismatch.
+unsafe trait OverrideAnyTypeId: Any {
+    fn overridden_type_id() -> TypeId;
+}
+
+#[allow(missing_debug_implementations)]
+#[doc(hidden)]
+#[unstable(feature = "drop_no_unwind_same_any_type_id", issue = "none")]
+#[repr(transparent)] // repr ≥ C to ensure same layout as `T`
+pub struct DropNoUnwindSameAnyTypeId<T>(ManuallyDrop<T>);
+
+#[unstable(feature = "drop_no_unwind_same_any_type_id", issue = "none")]
+impl<T> DropNoUnwindSameAnyTypeId<T> {
+    /// # Safety
+    ///
+    /// The value returned by `new()` must be immediately type erased to a
+    /// `Ptr<dyn Any>` so that no code sees the mismatch between static and
+    /// dynamic types.
+    #[doc(hidden)]
+    #[unstable(feature = "drop_no_unwind_same_any_type_id", issue = "none")]
+    pub unsafe fn new(value: T) -> Self {
+        Self(ManuallyDrop::new(value))
+    }
+}
+
+// SAFETY: from the safety preconditions of `DropNoUnwindSameAnyTypeId::new()`.
+unsafe impl<T: 'static> OverrideAnyTypeId for DropNoUnwindSameAnyTypeId<T> {
+    fn overridden_type_id() -> TypeId {
         TypeId::of::<T>()
     }
 }
 
+#[unstable(feature = "drop_no_unwind_same_any_type_id", issue = "none")]
+impl<T> Drop for DropNoUnwindSameAnyTypeId<T> {
+    fn drop(&mut self) {
+        struct AbortOnDrop {}
+
+        impl Drop for AbortOnDrop {
+            fn drop(&mut self) {
+                crate::intrinsics::abort();
+            }
+        }
+
+        let abort_on_drop_bomb = AbortOnDrop {};
+        // SAFETY: textbook ManuallyDrop.
+        unsafe {
+            ManuallyDrop::drop(&mut self.0);
+        }
+        mem::forget(abort_on_drop_bomb);
+    }
```

Comment:
I can see one more problem here: `catch_unwind` is often used to protect FFI boundary however I heard that drop bombs don't work there because unwinding implementation would attempt to process foreign stack frames and cause UB. So it seems this needs to be wrapped in `catch_unwind` instead and the result handled (by aborting on error).

Reply:
From my understanding, but I may be wrong, you can _start_ unwinding in an `extern "C"` function (or other similarly limitted FFI ABIs): the issue lies if the unwinding somehow makes it to the `extern "…" fn` boundary. So, assuming my understanding to be true, _both_ `catch_unwind` and an abort-on-unwind bomb ought to be fine, since they'd both be successful at preventing the unwind from escaping the `extern "…" fn`?

Reply:
This (daniel's post) matches my own understanding as well: the UB occurs when the unwind crosses an `extern "C"` boundary. So long as that does not happen, everything stays within Rust and works normally.

Reply:
I've refined my understanding a bit. The following is interpretations of disparate bits of platform-specific docs, but accurate to the best of my understanding.

<details><summary>Definitions</summary>
<dl>
<dt>MUST</dt><dd>describes an operation mandated by the runtime; not doing the operation fails to confirm to the ABI; exception: doing a MUST operation "after" an infinite loop or abort is sufficient</dd>
<dt>throw (an unwind)</dt><dd>register an unwind with the language-specific unwinding state and then raise it on the cross-language runtime; this cannot be done without knowing that language's unwind ABI refinement</dd>
<dt>raise (an unwind)</dt><dd>pass control flow and an unwind to the cross-language unwind runtime, starting unwinding the stack</dd>
<dt>suspend (an unwind)</dt><dd>stop unwinding frames to run a landing pad; the unwind stays owned by the unwind runtime; the unwind MUST be resumed</dd>
<dt>resume (an unwind)</dt><dd>after suspending an unwind, pass control flow back to the unwind runtime to continue unwinding the stack</dd>
<dt>catch (an unwind)</dt><dd>take control flow from the unwind runtime, ending the unwinding of the stack</dt>
<dt>landing pad</dt><dd>generated code which temporarily suspends an unwind, runs `finally` and other RAII destructors, and MUST resume the suspended unwind</dd>
<dt>catch handler</dt>a point at which the stack unwinding is terminated and the unwind is caught, passed back to normal control flow; if a foreign unwind is caught, its registered cleanup MUST be called before any other unwinding operations are performed on this thread</dt>
<dt>rethrow (an unwind)</dt><dd>throwing a previously caught unwind; if an unwind will be rethrown, the cleanup and setup for the unwind can be skipped, but only by refining the cross-language ABI to the language-specific unwinding ABI</dd>
<dt>foreign unwind</dt><dd>an unwind not produced by Rust but which conforms to the C-unwind ABI, allowing cross-language unwind interoperability</dd>
</dl>
</details>

0th: Across a C ABI boundary, no unwinds may happen;

1st: Per the C-unwind RFC as currently accepted, it is defined that (assuming all foreign code correctly conforms to the C-unwind ABI as defined-by-reference in Rust):

- Any unwind which would unwind over a C ABI will abort, though the exact timing of the abort is implementation defined[^c-abi-unwind-abort];
- Across a C-unwind ABI boundary:
  - A Rust unwind will unwind through foreign C-unwind-compatible frames (running landing pads);
  - A foreign unwind will unwind through Rust frames (running landing pads, aborting if `-Cpanic=abort`).

[^c-abi-unwind-abort]: Namely, this means that whether destructors of intermediate frames are run is implementation defined. Currently, a landing pad is registered which will abort when the unwind runtime hits it; this may still abort on some unwind runtimes if a catch handler is not present.

2nd: Per the C-unwind RFC as currently accepted, it is explicit UB (with intent to later define) if:

- A Rust unwind hits a catch handler in foreign code[^foreign-catches-rust];
- A foreign unwind hits a catch handler in Rust (`catch_unwind`)[^rust-catches-foreign];
- Throwing a Rust unwind during a foreign unwind or vice-versa[^nested].

[^foreign-catches-rust]: As currently implemented, this is a deterministic abort; defining this is trivial[^trivial], as its behavior is determined local to the Rust runtime by the provided cleanup callback.

[^rust-catches-foreign]: As currently implemented, this is a deterministic abort; defining this is trivial[^trivial], as its behavior is determined local to the Rust runtime as how it handles encountering a foreign unwind. If we do not abort, we MUST call cleanup on the foreign unwind and just supply `Box::new(ForeignUnwind{})` as the caught payload.

[^nested]: There's a lot of variables here. C++ `catch` is interesting[^catch]. Nesting rust in foreign can be defined to abort if we keep track of if a foreign unwind is in progress (i.e. update the unwind count when unwinds pass a C-unwind barrier). Nesting foreign in rust we have no control over.

[^trivial]: Trivial, presuming that the C-unwind ABI defines a slot for a cleanup callback; all currently known ABIs provide for cross language catching in this fashion, though it is not strictly *required* to allow foreign unwinds to pass over code if absolutely no unwind introspection is done.

[^catch]: I think C++ `catch` is always a catch handler for C++ unwinds and a plain landing pad for foreign unwinds (and if it's a foreign unwind the code MUST `throw;`). This decision is made knowing the exception on Itanium, and I think there is no actual difference on single-phase unwind mechanisms like SJLJ unwinds.

Reply:
(btw I hope you've backed this up somewhere or posted it in some more visible document, since it would be a pity for it to disappear underneath PR review comments 😄)

Reply:
Actively working on it and sort of fallen into a project-ffi-unwind membership because of it :)